Feat api.md for TDS-sensor library (#1)
amperka/AmperkaTDS: API.md

# AmperkaTDS API

## `class AmperkaTDS`

> Create an object of type `AmperkaTDS` to communicate with a particular [TDS-sensor](https://amperka.ru/product/troyka-tds-sensor).

Reviewer: We already discussed this in one of the previous requests. An object of type `AmperkaTDS` creates a class constructor, but here it is just a directive to the user.

> Create an object of type `AmperkaTDS` to communicate with a particular [TDS-sensor](https://amperka.ru/product/troyka-tds-sensor).

Author: > We already discussed this in one of the previous requests. An object of type `AmperkaTDS` creates a class constructor, but here it is just a directive to the user.

@Matthew3000, do you agree?

> Create an object of type `AmperkaTDS` to communicate with a particular [TDS-sensor](https://amperka.ru/product/troyka-tds-sensor).

Reviewer 2: Ok

### `AmperkaTDS(uint8_t pinS)`

Constructs a new `AmperkaTDS` object. The argument `pinS` the analog output signal. It is marked `S` on the TDS board. Use any analog GPIO pin of the controller.

### `void begin()`

Initializes the given interface, prepares the board for communication.
Call this method before interfacing with `AmperkaTDS`. For example, in your `setup()`.

> ### `void read()`

Reviewer: In this case, we can make the read method private, as it will only read the analog value from the sensor. And the calculation of the TDS and the EC can be done already in the methods getTDS and getEC.

Reviewer: :ok_hand:

Reads and stores the sensor data values: EC and TDS.

> ### `uint16_t getEC()`

Reviewer 3: Does it mutate the internal state? I guess not. If so, make it const. 

```suggestion
### `uint16_t getEC() const`
```

> Returns the stored electrical conductivity (EC) when the call method `read`. EC measurement is in microsiemens per centimeter and ranges from `0` to `2500` in `uS/cm`.

Reviewer 3: Non-code literals should not be enquoted:

```suggestion
Returns the stored electrical conductivity (EC) when the call method `read`. EC measurement is in microsiemens per centimeter and ranges from 0 to 2500 in uS/cm.
```

> ### `uint16_t getTDS()`

Reviewer 3: ```suggestion
### `uint16_t getTDS() const`
```

> Returns the stored total dissolved solids (TDS) when the call method `read`. TDS measurement is in parts per million and ranges from `0` to `1250` in `ppm`.

Reviewer 3: ```suggestion
Returns the stored total dissolved solids (TDS) when the call method `read`. TDS measurement is in parts per million and ranges from 0 to 1250 in ppm.
```

> ### `float getCalibrationFactor()`

Reviewer 3: Is it something that involves real communication with the sensor? I think a user should call this after putting the probe into a liquid with known parameters. Does he? If so, add a short description of the process and use verb _read_ in the signature:

```suggestion
### `float readCalibrationFactor()`
```

If it just returns a stored value, should be:

```suggestion
### `float getCalibrationFactor() const`
```

Author: @nkrkv You're right. I fixed this function and added a short description of the calibration.

> Returns the ratio between the known TDS value and the TDS measurement value. In the ideal case, the ratio is `1`. Returned value use in the function `setCalibrationFactor`. During the calibration, use a liquid buffer solution of known TDS value.

Reviewer 2: ```suggestion
Returns the ratio between the known TDS value and the TDS measurement value. In the ideal case, the ratio is `1`. Use the returned value in the function `setCalibrationFactor`. During the calibration, use a liquid buffer solution of known TDS value.
```

> _The measured value may differ from the actual value due to individual differences between different TDS sensors and differences in the control board. Therefore, to obtain a more accurate TDS value, calibration is required before measurement._

Reviewer 2: ```suggestion
_The measured value may differ from the actual value due to the individual differences between the different TDS sensors and differences in the control boards. Therefore, calibration is required before measurement, to obtain a more accurate TDS value, ._
```

Author: Done

### `void setCalibrationFactor(float calibrationFactor)`

> Corrects the difference between the known TDS value and the TDS measurement value. The argument `calibrationFactor` the ratio between the known TDS value and the TDS measurement value, which gets `getCalibrationFactor()` function.

Reviewer 2: ```suggestion
Corrects the difference between the known TDS value and the TDS measurement value. The argument `calibrationFactor` is the ratio between the known TDS value and the TDS measurement value, which gets `getCalibrationFactor()` function.
```

Author: Done

### `void setEnvironmentTemperature(float temperature)`

> Sets temperature from other climatic sensors to compensate TDS data. The argument `temperature` ambient temperature from other sensors. Range from `−25` to `50` °C.

Reviewer 2: ```suggestion
Sets temperature from other climatic sensors to compensate for TDS data error. The argument `temperature` is an ambient temperature from other sensors. Ranges from `−25` to `50` °C.
```

Author: Done

By default, the internal algorithm uses 25 °C temperature to compensate for environmental changes. Read temperature from other climatic sensors, e.g., [Troyka Meteo Sensor](https://amperka.ru/product/troyka-meteo-sensor) or [Sealed Temperature Sensor DS18B20](https://amperka.ru/product/sealed-temperature-sensor-ds18b20).

### `void setAnalogReadResolution(uint8_t bits = 10)`

> Sets resolution to output values, which used in calculate voltage on the sensor in function `read`. The argument `bits` is the current resolution analog to digital conversion (ADC) Arduino board. Ranges from `1` to `32` bits. If omitted, resolution output values are `10`. This function doesn't change the resolution Arduino board.

Reviewer 2: ```suggestion
Sets resolution to output values, that are used to calculate voltage on the sensor in function `read`. The argument `bits` is the current resolution analog to digital conversion (ADC) Arduino board. Ranges from `1` to `32` bits. If omitted, resolution output values are `10`. This function doesn't change the Arduino board resolution.
```

Author: Done

> Resolution ADC is set and changed standard Arduino function `analogReadResolution(uint8_t bits)`, but only some arduino boards. If you call the function `analogReadResolution (bits uint8_t)`, then call `setAnalogReadResolution (bits uint8_t)` with the same argument `bits`.

Reviewer 2: ```suggestion
Resolution ADC is set and changed with standard Arduino function `analogReadResolution(uint8_t bits)`, but only for some Arduino boards. If you call the function `analogReadResolution (bits uint8_t)`, then call `setAnalogReadResolution (bits uint8_t)` with the same argument `bits`.
```

Reviewer 2: Didn't quite understand the first sentence

Author: Corrected this sentence.

# AmperkaTDS API

## `class AmperkaTDS`

Create an object of type `AmperkaTDS` to communicate with a particular [TDS-sensor](https://amperka.ru/product/troyka-tds-sensor).

### `AmperkaTDS(uint8_t pinS)`

> Constructs a new `AmperkaTDS` object. The argument `pinS` the analog output signal. It is marked `S` on the TDS board. Use any analog GPIO pin of the controller.

Reviewer 2: ```suggestion
Constructs a new `AmperkaTDS` object. The argument `pinS` is the analog output signal. It is marked `S` on the TDS board. Use any analog GPIO pin of the controller.
```

### `void begin()`

Initializes the given interface, prepares the board for communication.
Call this method before interfacing with `AmperkaTDS`. For example, in your `setup`.

### `void read()`

Reads and stores the sensor data values: EC and TDS.

### `uint16_t getEC() const`

> Returns the stored electrical conductivity (EC) when the call method `read`. EC measurement is in microsiemens per centimeter and ranges from 0 to 2500 in uS/cm.

Reviewer 2: ```suggestion
Returns the stored electrical conductivity (EC) when the `read` method is called. EC measurement is in microsiemens per centimeter and ranges from 0 to 2500 in uS/cm.
```

### `uint16_t getTDS() const`

> Returns the stored total dissolved solids (TDS) when the call method `read`. TDS measurement is in parts per million and ranges from 0 to 1250 in ppm.

Reviewer 2: ```suggestion
Returns the stored total dissolved solids (TDS) when the `read` method is called. TDS measurement is in parts per million and ranges from 0 to 1250 in ppm.
```

### `void setCalibrationFactor(float calibrationFactor)`

> Corrects the difference between the known TDS value and the TDS measurement value. The argument `calibrationFactor` is the ratio between the known TDS value and the TDS measurement value, which returned `readCalibrationFactor` function.

Reviewer 2: ```suggestion
Corrects the difference between the known TDS value and the TDS measurement value. The argument `calibrationFactor` is the ratio between the known TDS value and the TDS measurement value, which is returned by `readCalibrationFactor` function.
```

### `float readCalibrationFactor(uint16_t calibrationTDS)`

> Calculate and returns the ratio between the known TDS value and the TDS measurement value. In the ideal case, the ratio is `1`. Use the returned value in the function `setCalibrationFactor`. The argument `calibrationTDS` the value liquid buffer solution with a known parameter TDS in ppm.

Reviewer 2: ```suggestion
Calculates and returns the ratio between the known TDS value and the TDS measurement value. In the ideal case, the ratio is `1`. Use the returned value in the function `setCalibrationFactor`. The argument `calibrationTDS` is the value of liquid buffer solution with a known parameter TDS in ppm.
```

> _The measured value TDS may differ from the actual value due to the individual differences between the different TDS sensors and differences in the control boards. Therefore, calibration is required before measurement, to obtain a more accurate TDS value. During the calibration, use a liquid buffer solution of known TDS value._

Reviewer 2: ```suggestion
_The measured TDS value may differ from the actual value due to the individual differences between the various TDS sensors and differences in the control boards. Therefore, calibration is required before measurement to obtain a more accurate TDS value. During the calibration, use a liquid buffer solution of known TDS value._
```

**Calibration Step**

- Insert the probe into the buffer solution of known TDS value.
- Call the function `readCalibrationFactor`, which returns the value `calibrationFactor`.
- Use the value `calibrationFactor` in the function `setCalibrationFactor`.

### `void setEnvironmentTemperature(float temperature)`

> Sets temperature from other climatic sensors to compensate for TDS data error. The argument `temperature` is an ambient temperature from other sensors. Ranges from `−25` to `50` °C.

Reviewer 4: "The argument `temperature` is an **ambient temperature** from other sensors."
Is `ambient temperature` property of liquid or air? If it is property of liquid, how will you use `Troyka Meteo Sensor`?

Reviewer 3: That’s a very interesting question which indeed needs to be reflected in docs. I’d like to see it too.

Author: Fixed.

Reviewer 4: ок

By default, the internal algorithm uses 25 °C temperature to compensate for environmental changes. Read temperature from other climatic sensors, e.g., [Troyka Meteo Sensor](https://amperka.ru/product/troyka-meteo-sensor) or [Sealed Temperature Sensor DS18B20](https://amperka.ru/product/sealed-temperature-sensor-ds18b20).

### `void setAnalogReadResolution(uint8_t bits = 10)`

> Sets resolution to output values, that are used to calculate voltage on the sensor in function `read`. The argument `bits` is the current resolution analog to digital conversion (ADC) Arduino board. Ranges from `1` to `32` bits. If omitted, resolution output values are `10`. This function doesn't change the Arduino board resolution.

Reviewer 2: ```suggestion
Sets the resolution to output values, that are used to calculate voltage on the sensor in function `read`. The argument `bits` is the current resolution analog to digital conversion (ADC) Arduino board. Ranges from `1` to `32` bits. If omitted, resolution output values are `10`. This function doesn't change the Arduino board resolution.
```
